Sources/MultiSets/CountedSet/CountedSet+Collection.swift

```diff
@@ -0,0 +1,105 @@
+//
+//===----------------------------------------------------------------------===//
+
+extension CountedSet: Collection {
+  /// The position of an element in a counted set.
+  @frozen
+  public struct Index: Comparable {
+    /// An index in the underlying dictionary storage of a counted set.
+    ///
+    /// This is used to distinguish between indices that point to elements with
+    /// different values.
+    @usableFromInline
+    let storageIndex: Dictionary<Element, UInt>.Index
+
+    /// The relative position of the element.
+    ///
+    /// This doesn't actually correspond to a distinct part of memory. Instead,
+    /// it is used to distinguish between indices that point to elements of the
+    /// same value.
+    ///
+    /// For example, the first index pointing to a value has an index of 0, the
+    /// second index pointing to that value will have an index of 1, and so on.
+    ///
+    /// When a counted set is subscripted with an index, the stored multiplicity
+    /// is retrieved using the `storageIndex` and compared with the `position`
+    /// to determine the index's validity.
+    @usableFromInline
+    let position: UInt
+
+    @inlinable
+    public static func < (
+      lhs: CountedSet<Element>.Index,
+      rhs: CountedSet<Element>.Index
+    ) -> Bool {
+      guard lhs.storageIndex != rhs.storageIndex else {
+        return lhs.storageIndex < rhs.storageIndex
+      }
+      return lhs.position < rhs.position
+    }
+
+    @usableFromInline
+    init(storageIndex: Dictionary<Element, UInt>.Index, position: UInt) {
+      self.storageIndex = storageIndex
+      self.position = position
+    }
+  }
+
+  @inlinable
+  public func index(after i: Index) -> Index {
+    guard i.position + 1 < rawValue[i.storageIndex].value else {
+      return Index(
+        storageIndex: rawValue.index(after: i.storageIndex),
+        position: 0
+      )
+    }
+
+    return Index(storageIndex: i.storageIndex, position: i.position + 1)
+  }
+
+  @inlinable
+  public subscript(position: Index) -> Element {
+    let keyPair = rawValue[position.storageIndex]
+    precondition(
+      position.position < keyPair.value,
+      "Attempting to access CountedSet elements using an invalid index"
+    )
+    return keyPair.key
+  }
+
+  ///  The number of elements in the set.
+  ///
+  /// - Complexity: O(*k*), where *k* is the number of unique elements in the
+  /// set.
+  @inlinable
+  public var count: Int {
+    Int(rawValue.values.reduce(.zero, +))
+  }
+
+  @inlinable
+  public var startIndex: Index {
+    Index(storageIndex: rawValue.startIndex, position: 0)
+  }
+
+  @inlinable
+  public var endIndex: Index {
+    Index(storageIndex: rawValue.endIndex, position: 0)
+  }
+
+  /// A value equal to the number of unique elements in the set.
+  ///
+  /// - Complexity: O(*k*), where *k* is the number of unique elements in the
+  /// set.
+  @inlinable
+  public var underestimatedCount: Int {
+    rawValue.count
```

Comment:
`underestimatedCount` has always been incredibly unclear to me, in terms of its intended role in `Collection`s. I don’t believe anything actually says it should be equivalent to `count` in such cases, so I figured it’d be a useful way to express the (less computationally expensive) number of unique elements.

If we change it to reference `count`, we should add another property with this value.

By the way, would it be possible to add a more efficient conditional implementation of Swift Algorithms’ [Unique](https://github.com/apple/swift-algorithms/blob/main/Guides/Unique.md) methods for `CountedSet`, in case anyone ends up using them together? I’m guessing the answer is no, but it’d be a nice-to-have.

Reply:
> `underestimatedCount` has always been incredibly unclear to me, in terms of its intended role in Collections. I don’t believe anything actually says it should be equivalent to count in such cases, so I figured it’d be a useful way to express the (less computationally expensive) number of unique elements.

This is a good idea in general, and `Collection` does allow this! `underestimatedCount` is typically only used in contexts that are generic over `Sequence`, but that doesn't mean it's a bad idea to make it diverge from `count` -- it can still be quite beneficial.

> By the way, would it be possible to add a more efficient conditional implementation of Swift Algorithms’ Unique methods for CountedSet, in case anyone ends up using them together? I’m guessing the answer is no, but it’d be a nice-to-have.

The way to do this right now would require Algorithms to import Collections, which would be a bad idea. 

However, Swift does have support for cross-module overlays, which are modules that automatically get loaded whenever some module A and module B are both imported to a source file. (Apple's SDKs use this to add, e.g., MapKit-related functionality to SwiftUI when a file imports both of these modules.) If this functionality was cleaned up & promoted to a full-blown language feature, including SwiftPM support for defining cross-import modules, then we could use that to provide uniquing overloads that take CountedSets.

Another alternative is to define a set of additional collection protocols in a separate package, and have swift-collections and swift-algorithms both depend on that. This would let swift-algorithms provide generic overloads for specific algorithms to speed them up when possible.

I don't think we need to do this for CountedMultiset right now, but it's something to think about for later. (FWIW, I think we'll eventually want to define some sort of a dictionary protocol at least, and possibly a UniqueCollection protocol.)

Reply:
Oh, the easiest way is to simply provide the uniquing methods directly on the counted multiset type, without importing Algorithms. This would work, but perhaps it would be even better to provide a `uniqueMembers` _view_ instead, exposing the underlying dictionary's `keys` view.

(I'm not sure we need to do this though -- we are already exposing the storage dictionary, after all, so folks can simply access `keys` directly.)